src/stan/variational/advi.hpp

```diff
@@ -522,7 +526,332 @@ class advi {
+        ret_mat(i) = std::log(-a(i) / ret_mat(i)) - ret_mat(i) - 1;
+      }
+      ret_mat = ret_mat.matrix();
+    };
+
+    auto adjust_k_wip = [&](double k_, double n_) mutable {
+      int a = 10, n_plus_a = n_ + a;
+      return k_ * n_ / n_plus_a + a * 0.5 / n_plus_a;
+    };
+    // end function definitions
+
+    if (sort_x) {
+      std::sort(x.data(), x.data() + x.size());
+    }
+
+    const int N = x.size(), prior = 3, M = min_grid_pts + 
+                                           std::floor(std::sqrt(N));
+    Eigen::VectorXd jj(M);
+    for(int i = 0; i < M; i++){
+      jj(i) = i + 1;  // seq_len(M) (cpp indexing)
+    }
+
+    Eigen::VectorXd theta(M);
+    double xstar = x((int)std::floor(N / 4 + 0.5));
+
+    theta = 1 / x(N-1) + (1 - (M / (jj.array() - 0.5)).sqrt()) / prior / xstar;
+    theta = theta.matrix();
+
+    Eigen::VectorXd l_theta(theta.size());
+    lx(theta, x, l_theta);
+    l_theta *= N;
+    
+    Eigen::VectorXd w_theta = Eigen::VectorXd::Ones(l_theta.size());
+    for(int i = 0; i < w_theta.size(); i++){
+      w_theta(i) = 1 / (l_theta.array() - l_theta(jj(i) - 1)).exp().sum();
+      // subtract 1 from jj(i) to convert to cpp index
+    }
+    double theta_hat = (theta.array() * w_theta.array()).sum();
+    k = (-theta_hat * x.array()).log1p().mean();
+    sigma = -k / theta_hat;
+
+    if(wip){
+      k = adjust_k_wip(k, N);
+    }
+
+    if(std::isnan(k)){
+      k = std::numeric_limits<double>::infinity();
+    }
+  }
+
+  /**
+   * RVI Diagnostics
+   * A sample function to show how compute_effective_sample_size() in
+   * stan/analyze/mcmc/compute_effective_sample_size.hpp would be used.
+   * 
+   * @param[in] samples An Eigen::VectorXd containing individual samples
+   * @return Calculated ESS
+   */
+
+
+  /**
+   * RVI Diagnostics
+   * A sample function to show how compute_potential_scale_reduction() in
+   * stan/analyze/mcmc/compute_potential_scale_reduction.hpp would be used to
+   * calculate Rhat values
+   * 
+   * @param[in] samples An Eigen::VectorXd containing individual samples
+   * @return Calculated Rhat
+   */
+  static double calculate_sample_rhat(const Eigen::VectorXd& samples){
+    std::vector<const double*> sample_ptr_arr = {samples.data()};
+    std::vector<size_t> size_arr = {(size_t)samples.size()};
+    return stan::analyze::compute_potential_scale_reduction(sample_ptr_arr,
+                                                            size_arr);
+  }
+ 
+  /**
+   * RVI Diagnostics
+   * Estimate the Effective Sample Size and Monte Carlo Standard Error of posterior samples where
+   * MCSE = sqrt( var(parmas) / ess)
+   * @param[in] samples An Eigen::VectorXd containing posterior samples @TODO rewrite from here
+   * @param[in] ess If specified, will be used as effective sample size instead
+   * of calling compute_effective_sample_size()
+   * 
+   * @return Calculated MCSE
+   */
+
+  inline double ESS_MCSE(double &ess, double &mcse,
```

Comment:
@hyunjimoon I'm sure the ess code is same as `compute_effective_sample_size()` in `src/stan/analyze/mcmc/compute_effective_sample_size.hpp`. Is there a reason why we can't use that functioni?

Reply:
Definitely use that function if it does what you want

Reply:
To compute the variance of samples combined from all chains, `chain_sq_mean` was needed so I added the following to the original code. There might be a better way I guess.
```
Eigen::VectorXd chain_sq_mean(num_chains);
chain_sq_mean(chain) = draw.array().square().mean();
*mcse = std::sqrt((chain_sq_mean.mean() - chain_mean.mean() * chain_mean.mean())/ess_val);
```